hir-transform/src/spill.rs

```diff
@@ -0,0 +1,996 @@
+#![allow(clippy::mutable_key_type)]
+use std::collections::{BTreeMap, VecDeque};
+
+use midenc_hir::{
+    self as hir,
+    adt::{SmallMap, SmallSet},
+    pass::{AnalysisManager, RewritePass, RewriteResult},
+    *,
+};
+use midenc_hir_analysis::{
+    spill::Placement, ControlFlowGraph, DominanceFrontier, DominatorTree, SpillAnalysis, Use, User,
+};
+use midenc_session::Session;
+use rustc_hash::FxHashSet;
+
+/// This pass inserts spills and reloads as computed by running a [SpillAnalysis] on the given
+/// function, recording materialized splits, spills, and reloads in the analysis results.
+///
+/// **IMPORTANT:** This pass is intended to be combined with the [RewriteSpills] pass when used
```

Comment:
Good suggestion! The main reason I didn't do that was twofold:

1. I would have to manually maintain the `AnalysisManager` state between the two passes (which honestly, isn't a huge deal, but mistakes there are much more problematic because using an outdated analysis for a rewrite (or codegen), will result in incorrect outputs. Currently that is handled automatically as part of the `RewriteSet` implementation (which is effectively a meta pass like what you describe, just more general).
2. With my other recent refactorings, the rewrite pass pipeline is now centrally managed in a single place, so it is no longer nearly as essential that passes avoid these kind of interdependencies. We construct a `RewriteSet` there, so it felt logical to place it there, rather than construct a nested `RewriteSet` somewhere else.

With those points in mind, do you still feel like it would be better to wrap it up in a single pass? I don't have strong feelings either way to be honest, so I'll err on the side of whatever makes it easier to follow what's going on.

Reply:
> 1. I would have to manually maintain the `AnalysisManager` state between the two passes (which honestly, isn't a huge deal, but mistakes there are much more problematic because using an outdated analysis for a rewrite (or codegen), will result in incorrect outputs. Currently that is handled automatically as part of the `RewriteSet` implementation (which is effectively a meta pass like what you describe, just more general).

Do you mean to invalidate the analyses between the two passes? It sounds like something could be encoded inside the "fused" Insert+Rewrite pass, right? So the tradeoff is break encoded analyses invalidation vs. the risk of using the passes incorrectly.

> 2. With my other recent refactorings, the rewrite pass pipeline is now centrally managed in a single place, so it is no longer nearly as essential that passes avoid these kind of interdependencies. We construct a `RewriteSet` there, so it felt logical to place it there, rather than construct a nested `RewriteSet` somewhere else.

This is a good point, but the presence of the warnings means that there is still a chance of using the pass incorrectly.
 
> With those points in mind, do you still feel like it would be better to wrap it up in a single pass? I don't have strong feelings either way to be honest, so I'll err on the side of whatever makes it easier to follow what's going on.

These are good points, but I think the benefits of the single pass are outweighed by the risks. I prefer the compiler stopping me from doing the wrong things rather than a comment.